sklearn/metrics/classification.py

```diff
@@ -457,8 +460,18 @@ def matthews_corrcoef(y_true, y_pred):
     lb.fit(np.hstack([y_true, y_pred]))
     y_true = lb.transform(y_true)
     y_pred = lb.transform(y_pred)
-    with np.errstate(invalid='ignore'):
-        mcc = np.corrcoef(y_true, y_pred)[0, 1]
+
+    u_yt = np.average(y_true, weights=sample_weight)
+    u_yp = np.average(y_pred, weights=sample_weight)
+
+    cov_ytyp = np.average((y_true - u_yt) * (y_pred - u_yp),
+                          weights=sample_weight)
+    cov_ytyt = np.average((y_true - u_yt) * (y_true - u_yt),
+                          weights=sample_weight)
+    cov_ypyp = np.average((y_pred - u_yp) * (y_pred - u_yp),
+                          weights=sample_weight)
```

Comment:
can you make it more efficient? ie avoid the numerous copies etc

or at least

(y_true - u_yt) * (y_true - u_yt) -> (y_true - u_yt) ** 2